air/src/range.rs

```diff
@@ -114,6 +164,60 @@ fn enforce_16bit<E: FieldElement>(frame: &EvaluationFrame<E>, result: &mut [E])
     constraint_count
 }
 
+// --- AUXILIARY COLUMNS (FOR MULTISET CHECKS) ----------------------------------------------------
+
+/// Ensures that the running product auxiliary column `p0` is correctly built up during the 8-bit
+/// section of the range check table and then correctly reduced in the 16-bit section of the table.
+///
+/// In the 8-bit section, when `t=0` the value of `z` is included in the running product at each
+/// step, and the constraint reduces to p0' = p0 * z.
+/// In the 16-bit section, when `t=1`, the running product is reduced by the difference between the
+/// current and next value, and the constraint reduces to p0' * (alpha + v' - v) = p0.
+fn enforce_running_product_p0<E, F>(
+    main_frame: &EvaluationFrame<F>,
+    aux_frame: &EvaluationFrame<E>,
+    alpha: E,
+    result: &mut [E],
+) -> usize
+where
+    F: FieldElement<BaseField = Felt>,
+    E: FieldElement<BaseField = Felt> + ExtensionOf<F>,
+{
+    let mut constraint_offset = 0;
+
+    let z = get_z(main_frame, alpha);
+    let t = main_frame.t().into();
+    let p0_term = aux_frame.p0() * (z - z * t + t);
+    let p0_next_term = aux_frame.p0_next()
+        * ((alpha + main_frame.v_next().into() - main_frame.v().into()) * t - t + E::ONE);
+    result[constraint_offset] = p0_next_term - p0_term;
+    constraint_offset += 1;
+
+    constraint_offset
+}
+
+/// Returns the value `z` which is included in the running product columns at each step. `z` causes
+/// the row's value to be included 0, 1, 2, or 4 times, according to the row's selector flags row.
+fn get_z<E, F>(main_frame: &EvaluationFrame<F>, alpha: E) -> E
+where
+    F: FieldElement<BaseField = Felt>,
+    E: FieldElement<BaseField = Felt> + ExtensionOf<F>,
+{
+    // Get the selectors and the value from the main frame.
+    let s0: E = main_frame.s0().into();
+    let s1: E = main_frame.s1().into();
+    let v: E = main_frame.v().into();
+
+    // Define the flags.
+    let f0: E = binary_not(s0) * binary_not(s1);
+    let f1: E = s0 * binary_not(s1);
+    let f2: E = binary_not(s0) * s1;
+    let f3: E = s0 * s1;
+
+    // Compute z.
+    f3 * (v + alpha).square().square() + f2 * (v + alpha).square() + f1 * (v + alpha) + f0
```

Comment:
The compiler might be able to optimize this, but I wonder if it would be better to compute this reusing previously computed values. For example, something like this:
```Rust
let v_alpha = v + alpha;
let v_alpha2 = v_alpha.square();
let v_alpha4 = v_alpha2.square();
f3 * v_alpha4 + f2 * v_alpha2 + f1 * v_alpha + f0
```

Reply:
Yeah, good point, thanks. Updated!